libs/pika/async_mpi/src/mpi_polling.cpp

```diff
@@ -53,7 +61,23 @@ namespace pika::mpi::experimental {
         /// Queries an environment variable to get/override a default value for
         /// the number of messages allowed 'in flight' before we throttle a
         /// thread trying to send more data
-        size_t get_throttling_default();
+        std::uint32_t get_throttling_default();
+
+        // -----------------------------------------------------------------
+        /// To enable independent throttling of sends/receives/other
+        /// we maintina several "queues" which have their own condition
+        /// variables for suspension
+        struct mpi_stream
+        {
+            mutex_type throttling_mtx_;
+            pika::lcos::local::condition_variable throttling_cond_;
+            std::atomic<std::uint32_t> in_flight_{0};
+            std::uint32_t limit_{get_throttling_default()};
+            std::uint32_t index;
+        };
+
+        using mpi_callback_queue_tuple =
+            std::tuple<request_callback_function_type, mpi_stream*>;
```

Comment:
Seeing this again this would probably be better off in a struct, but that can be separately.